src/crimson/net/ProtocolV1.cc

```diff
@@ -0,0 +1,851 @@
+    return { reinterpret_cast<const char*>(&value), sizeof(value) };
+}
+
+// store the banner in a non-const string for buffer::create_static()
+char banner[] = CEPH_BANNER;
+constexpr size_t banner_size = sizeof(CEPH_BANNER)-1;
+
+constexpr size_t client_header_size = banner_size + sizeof(ceph_entity_addr);
+constexpr size_t server_header_size = banner_size + 2 * sizeof(ceph_entity_addr);
+
+// check that the buffer starts with a valid banner without requiring it to
+// be contiguous in memory
+void validate_banner(bufferlist::const_iterator& p)
+{
+  auto b = std::cbegin(banner);
+  auto end = b + banner_size;
+  while (b != end) {
+    const char *buf{nullptr};
+    auto remaining = std::distance(b, end);
+    auto len = p.get_ptr_and_advance(remaining, &buf);
+    if (!std::equal(buf, buf + len, b)) {
+      throw std::system_error(
+          make_error_code(ceph::net::error::bad_connect_banner));
+    }
+    b += len;
+  }
+}
+
+// make sure that we agree with the peer about its address
+void validate_peer_addr(const entity_addr_t& addr,
+                        const entity_addr_t& expected)
+{
+  if (addr == expected) {
+    return;
+  }
+  // ok if server bound anonymously, as long as port/nonce match
+  if (addr.is_blank_ip() &&
+      addr.get_port() == expected.get_port() &&
+      addr.get_nonce() == expected.get_nonce()) {
+    return;
+  } else {
+    throw std::system_error(
+        make_error_code(ceph::net::error::bad_peer_address));
+  }
+}
+
+// return a static bufferptr to the given object
+template <typename T>
+bufferptr create_static(T& obj)
+{
+  return buffer::create_static(sizeof(obj), reinterpret_cast<char*>(&obj));
+}
+
+uint32_t get_proto_version(entity_type_t peer_type, bool connect)
+{
+  constexpr entity_type_t my_type = CEPH_ENTITY_TYPE_OSD;
+  // see also OSD.h, unlike other connection of simple/async messenger,
+  // crimson msgr is only used by osd
+  constexpr uint32_t CEPH_OSD_PROTOCOL = 10;
+  if (peer_type == my_type) {
+    // internal
+    return CEPH_OSD_PROTOCOL;
+  } else {
+    // public
+    switch (connect ? peer_type : my_type) {
+      case CEPH_ENTITY_TYPE_OSD: return CEPH_OSDC_PROTOCOL;
+      case CEPH_ENTITY_TYPE_MDS: return CEPH_MDSC_PROTOCOL;
+      case CEPH_ENTITY_TYPE_MON: return CEPH_MONC_PROTOCOL;
+      default: return 0;
+    }
+  }
+}
+
+void discard_up_to(std::queue<MessageRef>* queue,
+                   ceph::net::seq_num_t seq)
+{
+  while (!queue->empty() &&
+         queue->front()->get_seq() < seq) {
+    queue->pop();
+  }
+}
+
+} // namespace anonymous
+
+namespace ceph::net {
+
+ProtocolV1::ProtocolV1(Dispatcher& dispatcher,
+                       SocketConnection& conn,
+                       SocketMessenger& messenger)
+  : Protocol(1, dispatcher, conn), messenger{messenger} {}
+
+ProtocolV1::~ProtocolV1() {}
+
+// connecting state
+
+void ProtocolV1::reset_session()
+{
+  decltype(conn.out_q){}.swap(conn.out_q);
```

Comment:
I think this is because `std::queue` lacks `::clear()`. However, maybe we could just assign from an empty instance (`connt.out_q = std::queue{}`)?

Reply:
i changed it to
```c++
connt.out_q = {};
```